File: crm_backend/app/model/order.go
// app/models/inventory.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type OrderStatusType string

const (
	OrderStatusPending   StatusType = "Pending"
	OrderStatusConfirmed StatusType = "Confirmed"
	OrderStatusCancelled StatusType = "Cancelled"
)

type Order struct {
	gorm.Model
	Status        string     `gorm:"type:enum('Pending', 'Confirmed', 'Cancelled');default:'Pending'"`
	Price         int        `gorm:"not null"`
	ConfirmedDate *time.Time // Date when the GRN is confirmed (can be null)
	CancelledDate *time.Time // Date when the GRN is confirmed (can be null)

	CreatedByID uint `json:"-"`
	CreatedBy   User `gorm:"foreignKey:CreatedByID" json:"created_by"`

	CreatedForID uint `json:"-"`
	CreatedFor   User `gorm:"foreignKey:CreatedForID" json:"created_for"`

	ConfirmedByID *uint // User ID who confirmed the GRN (can be null)
	ConfirmedBy   *User `gorm:"foreignKey:ConfirmedByID" json:"confirmed_by"`

	CancelledByID *uint //
	CancelledBy   *User `gorm:"foreignKey:CancelledByID" json:"cancelled_by"`

	OrderProducts []OrderProduct `gorm:"foreignKey:OrderID"`
}

type OrderProduct struct {
	gorm.Model
	OrderID uint  `gorm:"not null;uniqueIndex:idx_order_product_size"` // Foreign key to the GRN
	Order   Order `gorm:"foreignKey:OrderID"`

	ProductID uint    `gorm:"not null;uniqueIndex:idx_order_product_size"` // Foreign key to the Product
	Product   Product `gorm:"foreignKey:ProductID"`

	SizeVariant1ID uint        `gorm:"not null;uniqueIndex:idx_order_product_size"`
	SizeVariant1   SizeVariant `gorm:"foreignKey:SizeVariant1ID" json:"size_variant_1"`

	SizeVariant2ID *uint
	SizeVariant2   *SizeVariant `gorm:"foreignKey:SizeVariant2ID" json:"size_variant_2"`

	SizeVariant3ID *uint
	SizeVariant3   *SizeVariant `gorm:"foreignKey:SizeVariant3ID" json:"size_variant_3"`

	Quantity int `gorm:"not null"` // Quantity of this specific product in this GRN
	Price    int `gorm:"not null"`
}

File: crm_ui/src/model/order.ts
import Product, { SizeVariant, SpecificSizeVariant } from "./product";
import User from "./user";

// api request
export interface PlaceOrderRequest {
	order: CartItem[],
	total_price: number;
}

export interface PlaceOrderAdminRequest {
	products: {
		product_id: number;
		size_variants: SpecificSizeVariant[];
		price: number;
	}[],
	created_for_id: number;
	total_price: number;
}


// api response
export interface GetOrdersResponse {
	"page": number,
	"page_size": number,
	"total_items": number,
	"total_pages": number,
	"orders": Order[];
}

export interface AddOrderResponse {
	status: string;
	order: Order;
}

// model
export class CartItem {
	product_id: number = 0;
	sets: SetForCat[] = [];
}

export class SetForCat {
	size1: SizeVariant;
	size2: SizeVariant;
	size3: SizeVariant;
	quantity: number;
	price: number = 0;

	constructor(size1: SizeVariant, size2: SizeVariant, size3: SizeVariant, quantity: number) {
		this.size1 = size1;
		this.size2 = size2;
		this.size3 = size3;
		this.quantity = quantity;
	}
}


export default class Order {
	ID: number;
	CreatedAt: string;
	UpdatedAt: string;
	Status: string;
	Price: number;
	created_by: User;
	created_for: User;
	confirmed_by: User | null = null;
	cancelled_by: User | null = null;
	OrderProducts: OrderProducts[] = [];
	constructor(ID: number, CreatedAt: string, UpdatedAt: string, Status: string, Price: number, created_by: User, created_for: User) {
		this.ID = ID;
		this.CreatedAt = CreatedAt;
		this.UpdatedAt = UpdatedAt;
		this.Status = Status;
		this.Price = Price;
		this.created_by = created_by;
		this.created_for = created_for;
	}
}

export class OrderProducts {
	ID: number;
	CreatedAt: string;
	UpdatedAt: string;
	ProductID: number;
	Product: Product;
	size_variant_1: SpecificSizeVariant;
	size_variant_2: SpecificSizeVariant | null;
	size_variant_3: SpecificSizeVariant | null;
	Quantity: number;
	Price: number;

	constructor(ID: number, CreatedAt: string, UpdatedAt: string, ProductID: number, Product: Product, size_variant_1: SpecificSizeVariant, size_variant_2: SpecificSizeVariant, size_variant_3: SpecificSizeVariant, Quantity: number, Price: number,) {
		this.ID = ID;
		this.CreatedAt = CreatedAt;
		this.UpdatedAt = UpdatedAt;
		this.ProductID = ProductID;
		this.Product = Product;
		this.size_variant_1 = size_variant_1;
		this.size_variant_2 = size_variant_2;
		this.size_variant_3 = size_variant_3;
		this.Quantity = Quantity;
		this.Price = Price;
	}
}